x-pack/filebeat/input/s3/s3_integration_test.go

```diff
@@ -150,103 +173,68 @@ func defaultTestConfig() *common.Config {
 	})
 }
 
-func runTest(t *testing.T, cfg *common.Config, run func(t *testing.T, input *s3Input, out *stubOutleter)) {
-	// Simulate input.Context from Filebeat input runner.
-	inputCtx := newInputContext()
-	defer close(inputCtx.Done)
-
-	// Stub outlet for receiving events generated by the input.
-	eventOutlet := newStubOutlet()
-	defer eventOutlet.Close()
-
-	connector := channel.ConnectorFunc(func(_ *common.Config, _ beat.ClientConfig) (channel.Outleter, error) {
-		return eventOutlet, nil
-	})
-
-	in, err := NewInput(cfg, connector, inputCtx)
+func runTest(t *testing.T, cfg *common.Config, run func(t *testing.T, input *s3Collector, receiver *eventReceiver)) {
+	plugin := Plugin()
+	inp, err := plugin.Manager.Create(cfg)
 	if err != nil {
 		t.Fatal(err)
 	}
-	s3Input := in.(*s3Input)
-	defer s3Input.Stop()
 
-	run(t, s3Input, eventOutlet)
-}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-func newInputContext() input.Context {
-	return input.Context{
-		Done: make(chan struct{}),
+	client := pubtest.NewChanClient(0)
+	receiver := &eventReceiver{client.Channel}
+	pipeline := pubtest.ConstClient(client)
+	collector, err := inp.(*s3Input).createCollector(v2.Context{
+		Logger:      logp.NewLogger("test"),
+		Cancelation: ctx,
+	}, pipeline)
+	if err != nil {
+		t.Fatal(err)
 	}
-}
 
-type stubOutleter struct {
-	sync.Mutex
-	cond   *sync.Cond
-	done   bool
-	Events []beat.Event
+	run(t, collector, receiver)
 }
```

Comment:
The way runTest is used, consider to provide a setup instead (having no indirection via functions can make it easier to read the code).

e.g.

```
func setupInput(t *testing.T, ...) (*s3Collector, ...) {
   inp, err := Plugin().Manager.Create(cfg)
   if err != nil {
     t.Fatal(err)
   }

   ctx, cancel := context.WithCancel(context.Background())
   t.Cleanup(cancel)  // automatically run when test returns

   ...
}
```

Reply:
Done! Thank you for the suggestion!